### PR TITLE
Made OOP to handle initial solution case better and made it to have better perf.

### PR DESCRIPTION
**Pull request by heejaechang**
**Customer scenario**

Customer has either code lens/FSA on, and just opened new solution. and in some cases, VS sync some data twice.

**Bugs this fixes:**

This doesn't have an associate bug, but one of PRs to improve OOP perf on the initial solution synchronization.

**Workarounds, if any**

There is no workaround. before the very first primary solution synchronization, all solution treated as temporary solution and not reused to incrementally update the solution.

**Risk**

Low. There should be no functional behavior change.

**Performance impact**

Should make initial solution perf better in OOP 

**Is this a regression from a previous update?**

No

**Root cause analysis:**

This is not regression. it was just the way it is designed. now I am trying to make this particular case to have better perf by not making all solution a temporary solution before the very first primary workspace synchronization.

**How was the bug found?**

While doing FAR/GoTo OOP work.

### CODE REVIEW

**Comment by heejaechang** [comment]
@CyrusNajmabadi want to take a look? have some changes from #19277 though.

**Comment by heejaechang** [comment]
can I get some code review? @CyrusNajmabadi @mavasani @dotnet/roslyn-analysis

File: src/VisualStudio/Core/Next/Remote/JsonRpcSession.cs [hunk 1]
```diff
@@ -89,10 +89,10 @@ internal class JsonRpcSession : RemoteHostClient.Session
             // all roslyn remote service must based on ServiceHubServiceBase which implements Initialize method
             if (_snapshotClientOpt != null)
             {
-                await _snapshotClientOpt.InvokeAsync(WellKnownServiceHubServices.ServiceHubServiceBase_Initialize, _currentSessionId, PinnedScopeOpt.SolutionChecksum).ConfigureAwait(false);
+                await _snapshotClientOpt.InvokeAsync(WellKnownServiceHubServices.ServiceHubServiceBase_Initialize, _currentSessionId, PinnedScopeOpt.Primary, PinnedScopeOpt.SolutionChecksum).ConfigureAwait(false);
             }
 
-            await _serviceClient.InvokeAsync(WellKnownServiceHubServices.ServiceHubServiceBase_Initialize, _currentSessionId, PinnedScopeOpt?.SolutionChecksum).ConfigureAwait(false);
+            await _serviceClient.InvokeAsync(WellKnownServiceHubServices.ServiceHubServiceBase_Initialize, _currentSessionId, PinnedScopeOpt?.Primary ?? false, PinnedScopeOpt?.SolutionChecksum).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Next/Remote/JsonRpcSession.cs]: can you comment this to explain what's going on?

File: src/Workspaces/Core/Portable/Execution/AssetStorages.cs [hunk 2]
```diff
@@ -156,13 +157,21 @@ public RemotableData GetRemotableData(PinnedRemotableDataScope scope, Checksum c
         public void RegisterSnapshot(PinnedRemotableDataScope snapshot, AssetStorages.Storage storage)
         {
             // duplicates are not allowed, there can be multiple snapshots to same solution, so no ref counting.
-            Contract.ThrowIfFalse(_storages.TryAdd(snapshot, storage));
+            if (!_storages.TryAdd(snapshot, storage))
+            {
+                // this should make failure more explicit
+                FailFast.OnFatalException(new Exception("who is adding same snapshot?"));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/AssetStorages.cs]: adding @tmeschter .  Will this get a stack trace appropriately?

File: src/Workspaces/Core/Portable/Execution/PinnedRemotableDataScope.cs [hunk 3]
```diff
@@ -33,6 +33,8 @@ internal sealed class PinnedRemotableDataScope : IDisposable
             _storages.RegisterSnapshot(this, storage);
         }
 
+        public bool Primary => _storage.SolutionState.BranchId == Workspace.PrimaryBranchId;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/PinnedRemotableDataScope.cs]: IsPrimaryBranchId.

File: src/Workspaces/Remote/Core/Services/SolutionCreator.cs [hunk 4]
```diff
@@ -613,5 +612,64 @@ private ImmutableArray<string> GetStrongNameKeyPaths(ProjectInfo.ProjectAttribut
...
+                sb.AppendLine($"{kv.Key.ToString()}, {kv.Value.ToString()}");
+            }
+
+            Logger.Log(FunctionId.SolutionCreator_AssetDifferences, sb.ToString());
+#endif
+
+            return;
+        }
+
+        private async Task RemoveDuplicateChecksumsAsync(Checksum givenSolutionChecksum, Dictionary<Checksum, object> map)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/Services/SolutionCreator.cs]: [Condition(debug)]

File: src/Workspaces/Remote/Core/Services/SolutionService.cs [hunk 5]
```diff
@@ -29,7 +29,15 @@ public SolutionService(AssetService assetService)
             _assetService = assetService;
         }
 
-        public async Task<Solution> GetSolutionAsync(Checksum solutionChecksum, CancellationToken cancellationToken)
+        public Task<Solution> GetSolutionAsync(Checksum solutionChecksum, CancellationToken cancellationToken)
+        {
+            return GetSolutionAsync(solutionChecksum, primary: false, cancellationToken: cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/Services/SolutionService.cs]: doc why primary:false.

File: src/Workspaces/Remote/Core/Services/SolutionService.cs [hunk 6]
```diff
@@ -46,67 +54,63 @@ public SolutionService(AssetService assetService)
...
                 s_lastSolution = Tuple.Create(solutionChecksum, solution);
 
                 return solution;
             }
         }
 
+        /// <summary>
+        /// This should be only consumed by engine, not by user.
+        /// </summary>
         public async Task UpdatePrimaryWorkspaceAsync(Checksum solutionChecksum, CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/Services/SolutionService.cs]: Add _DoNotCallDirectly.

File: src/Workspaces/Remote/Core/Services/SolutionService.cs [hunk 7]
```diff
@@ -46,67 +54,63 @@ public SolutionService(AssetService assetService)
...
 
-        private async Task<Solution> CreateSolution_NoLockAsync(Checksum solutionChecksum, Solution baseSolution, CancellationToken cancellationToken)
+        private async Task<Solution> CreateSolution_NoLockAsync(Checksum solutionChecksum, bool primary, Solution baseSolution, CancellationToken cancellationToken)
         {
             var updater = new SolutionCreator(_assetService, baseSolution, cancellationToken);
 
             if (await updater.IsIncrementalUpdateAsync(solutionChecksum).ConfigureAwait(false))
             {
                 // solution has updated
+                if (primary)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/Services/SolutionService.cs]: document all of this.

File: src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs [hunk 8]
```diff
@@ -28,6 +28,7 @@ internal abstract class ServiceHubServiceBase : IDisposable
         protected readonly CancellationToken CancellationToken;
 
         private int _sessionId;
+        private bool _primary;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs]: document what this means.

File: src/Workspaces/Core/Portable/Execution/AssetStorages.cs (hunk 2 above)
Comment (tmeschter) [inline src/Workspaces/Core/Portable/Execution/AssetStorages.cs]: Ultimately this ends up calling `Microsoft.CodeAnalysis.Environment.FailFast()` which only uses the exception for its `Message` property. So in practice it should be ok.

File: src/Workspaces/Remote/Core/Services/SolutionCreator.cs (hunk 4 above)
Comment (heejaechang) [inline src/Workspaces/Remote/Core/Services/SolutionCreator.cs]: can't since it is async method.

**Comment by heejaechang** [comment]
@CyrusNajmabadi  can you take a look?

**Comment by CyrusNajmabadi** [comment]
Looking.

File: src/VisualStudio/Core/Next/Remote/JsonRpcSession.cs [hunk 9]
```diff
@@ -86,13 +86,17 @@ internal class JsonRpcSession : RemoteHostClient.Session
...
         {
-            // all roslyn remote service must based on ServiceHubServiceBase which implements Initialize method
+            // All roslyn remote service must based on ServiceHubServiceBase which implements Initialize method
+            // This will set this session's solution and whether that solution is for primary branch or not
+            var primaryBranch = PinnedScopeOpt?.ForPrimaryBranch ?? false;
+            var solutionChecksum = PinnedScopeOpt?.SolutionChecksum;
+
             if (_snapshotClientOpt != null)
             {
-                await _snapshotClientOpt.InvokeAsync(WellKnownServiceHubServices.ServiceHubServiceBase_Initialize, _currentSessionId, PinnedScopeOpt.SolutionChecksum).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Next/Remote/JsonRpcSession.cs]: is it ok that solutionChecksum might be null here?

File: src/Workspaces/Core/Portable/Execution/PinnedRemotableDataScope.cs [hunk 10]
```diff
@@ -33,6 +33,11 @@ internal sealed class PinnedRemotableDataScope : IDisposable
             _storages.RegisterSnapshot(this, storage);
         }
 
+        /// <summary>
+        /// This indicates whether this scope is for primary branch or not (not forked solution)
+        /// </summary>
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Execution/PinnedRemotableDataScope.cs]: comments shoudl explain why this is important.  What will consumers use this for?

File: src/Workspaces/Remote/Core/Services/SolutionCreator.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/Services/SolutionCreator.cs]: Async methods can't be conditioned? WTF.

File: src/VisualStudio/Core/Next/Remote/JsonRpcSession.cs (hunk 9 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Next/Remote/JsonRpcSession.cs]: PinnedScope.SolutionChecksum can't be null. let me add contract.

File: src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs [hunk 11]
```diff
@@ -27,8 +26,22 @@ internal abstract class ServiceHubServiceBase : IDisposable
...
         protected readonly CancellationToken CancellationToken;
 
+        /// <summary>
+        /// Session Id of this service. caller and callee share this id which one
+        /// can use to find matching caller and callee when debugging or logging
+        /// </summary>
         private int _sessionId;
+
+        /// <summary>
+        /// Mark whether the solution checksum it got is for primary branch or not
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs]: we need a comment somewhere to explain why this is impoirtant.  why do we care if it's from the primary branch or not?  How do we treat things differently if it's from the primary branch or not.

Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs]: how does this impact features?  etc. etc.

File: src/Workspaces/Core/Portable/Execution/PinnedRemotableDataScope.cs (hunk 10 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/PinnedRemotableDataScope.cs]: it tells whether the solution this checksum is created from is from primary branch or not. it depends whether the consumer care about the solution is from primary branch or not. for example, OOP framework care.

File: src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs (hunk 11 above)
Comment (heejaechang) [inline src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs]: that comment is in SolutionService where the flag is consumed. this is all plumbing.

**Comment by heejaechang** [comment]
@CyrusNajmabadi is comment all you want?

**Comment by CyrusNajmabadi** [comment]
Basically yes.  Imagine that you're not here and someone needs to figure htis out :)

File: src/VisualStudio/Core/Next/Remote/JsonRpcSession.cs (hunk 9 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Next/Remote/JsonRpcSession.cs]: added contract on PinnedScope itself on solution checksum

**Comment by heejaechang** [comment]
@CyrusNajmabadi can you take a look? updated comments

**Comment by CyrusNajmabadi** [comment]
will do.

File: src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs [hunk 12]
```diff
@@ -27,8 +26,25 @@ internal abstract class ServiceHubServiceBase : IDisposable
...
+        /// Session Id of this service. caller and callee share this id which one
+        /// can use to find matching caller and callee when debugging or logging
+        /// </summary>
         private int _sessionId;
+
+        /// <summary>
+        /// Mark whether the solution checksum it got is for primary branch or not 
+        /// 
+        /// this flag will be passed down to solution controller to help
+        /// solution service's cache policy. for more detail, see <see cref="SolutionService"/>
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs]: i looked in SolutionService and i couldn't see clear documentation explaining this.  Sorry, i think we need a very clear explanation at some point stating how Solutions are synced, and how primary/non-primary matters.

Comment (heejaechang) [inline src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs]: comments in CreateSolution_NoLockAsync not enough?

Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs]: > comments in CreateSolution_NoLockAsync not enough?

I don't believe so.  I'm looking for like a top level comment on some class or method that explains the overall design here.  

i'm looking for: 

How does VS+OOP sync up a solution.  And, specifically, how that varies in different circumstances (i.e. if it's the priary solution, here's how it works.  if not primary, here's what's different).  And why we want this.  i.e. why this gives a performant system.  

Imagine i'm new and i'm trying to figure out why you did things the way you did.

Comment (heejaechang) [inline src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs]: sure no problem. I will add those.

**Comment by CyrusNajmabadi** [comment]
Want me to review?

**Comment by CyrusNajmabadi** [comment]
Thanks!
👍

**Comment by heejaechang** [comment]
@MattGertz can I get approval for master?

**Pull request merged**